rotkehlchen/exchanges/iconomi.py

```diff
@@ -0,0 +1,260 @@
+        signature = hmac.new(
+            self.secret,
+            signed_data,
+            hashlib.sha512,
+        )
+        self.session.headers.update({
+            'ICN-SIGN': base64.b64encode(signature.digest()).decode(),
+        })
+
+    def _api_query(
+            self,
+            verb: str,
+            path: str,
+            options: Optional[Dict] = None,
+    ) -> Dict:
+        """
+        Queries ICONOMI with the given verb for the given path and options
+        """
+        assert verb in ('get', 'post', 'push'), (
+            'Given verb {} is not a valid HTTP verb'.format(verb)
+        )
+
+        request_path_no_args = '/v1/' + path
+
+        data = ''
+        if not options:
+            request_path = request_path_no_args
+        else:
+            request_path = request_path_no_args + '?' + urlencode(options)
+
+        timestamp = str(int(time.time() * 1000))
+        request_url = self.uri + request_path
+
+        self._generate_signature(
+            request_type=verb.upper(),
+            request_path=request_path_no_args,
+            timestamp=timestamp,
+        )
+
+        self.session.headers.update({
+            'ICN-TIMESTAMP': timestamp,
+        })
+        if data != '':
+            self.session.headers.update({
+                'Content-Type': 'application/json',
+                'Content-Length': str(len(data)),
+            })
+
+        log.debug('ICONOMI API Query', verb=verb, request_url=request_url)
+
+        try:
+            response = getattr(self.session, verb)(request_url, data=data, timeout=30)
+        except ReadTimeout as exc:
+            raise RemoteError('Read timeout') from exc
+
+        try:
+            json_ret = rlk_jsonloads(response.text)
+        except JSONDecodeError as exc:
+            raise RemoteError('ICONOMI returned invalid JSON response') from exc
+
+        if response.status_code not in (200, 201):
+            if isinstance(json_ret, dict) and 'message' in json_ret:
+                raise RemoteError(json_ret['message'])
+
+            raise RemoteError(
+                'ICONOMI api request for {} failed with HTTP status code {}'.format(
+                    response.url,
+                    response.status_code,
+                ),
+            )
+
+        if not isinstance(json_ret, dict):
+            raise RemoteError('ICONOMI returned invalid non-dict response')
+
+        return json_ret
+
+    def validate_api_key(self) -> Tuple[bool, str]:
+        "Validates that the ICONOMI API key is good for usage in Rotki"
+
+        try:
+            self._api_query('get', 'user/balance')
+            return True, ""
+
+        except RemoteError:
+            return False, 'Provided API Key is invalid'
+
+    def query_balances(self, **kwargs: Any) -> Tuple[Optional[Dict[Asset, Dict[str, Any]]], str]:
+        balances = {}
+        resp_info = self._api_query('get', 'user/balance')
```

Comment:
Here, just like in other exchange `query_balances()` put this in try/except and catch `RemoteError`. If there is an error log it and return `None, error_msg`.

Example:

https://github.com/rotki/rotki/blob/4500381d5c527d255256f1a32d8261722ce028c2/rotkehlchen/exchanges/bittrex.py#L306-L315